sdk/identity/Azure.Identity/src/VisualStudioCodeCredential.cs

```diff
@@ -0,0 +1,153 @@
+            : this(tenantId, CredentialPipeline.GetInstance(options), fileSystem, vscAdapter)
+        {
+        }
+
+        internal VisualStudioCodeCredential(string tenantId, CredentialPipeline pipeline, IFileSystemService fileSystem, IVisualStudioCodeAdapter vscAdapter)
+            : this(tenantId, pipeline, pipeline.CreateMsalPublicClient(ClientId), fileSystem, vscAdapter)
+        {
+        }
+
+        internal VisualStudioCodeCredential(string tenantId, CredentialPipeline pipeline, MsalPublicClient client, IFileSystemService fileSystem, IVisualStudioCodeAdapter vscAdapter)
+        {
+            _tenantId = tenantId ?? "common";
+            _pipeline = pipeline;
+            _client = client;
+            _fileSystem = fileSystem ?? FileSystemService.Default;
+            _vscAdapter = vscAdapter ?? GetVscAdapter();
+        }
+
+        /// <inheritdoc />
+        public override async ValueTask<AccessToken> GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)
+            => await GetTokenImplAsync(requestContext, true, cancellationToken).ConfigureAwait(false);
+
+        /// <inheritdoc />
+        public override AccessToken GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken)
+            => GetTokenImplAsync(requestContext, false, cancellationToken).EnsureCompleted();
+
+        private async ValueTask<AccessToken> GetTokenImplAsync(TokenRequestContext requestContext, bool async, CancellationToken cancellationToken)
+        {
+            using CredentialDiagnosticScope scope = _pipeline.StartGetTokenScope("VisualStudioCodeCredential.GetToken", requestContext);
+
+            try
+            {
+                GetUserSettings(out var tenant, out var environmentName);
+
+                var cloudInstance = GetAzureCloudInstance(environmentName);
+                var storedCredentials = _vscAdapter.GetCredentials(CredentialsSection, environmentName);
+
+                if (!IsBase64String(storedCredentials))
+                {
+                    throw new CredentialUnavailableException("Need to re-authenticate user in VSCode Azure Account.");
+                }
+
+                var result = await _client.AcquireTokenWithDeviceCodeAsync(requestContext.Scopes, storedCredentials, cloudInstance, tenant, async, cancellationToken).ConfigureAwait(false);
+                return scope.Succeeded(new AccessToken(result.AccessToken, result.ExpiresOn));
+            }
+            catch (OperationCanceledException e)
+            {
+                scope.Failed(e);
+                throw;
+            }
+            catch (Exception e)
+            {
+                throw scope.FailAndWrap(e);
+            }
+        }
+
+        private static bool IsBase64String(string str)
+            => str.Select(t => (uint)t).All(ch => ch >= '0' && ch <= '9' || ch >= 'A' && ch <= 'Z' || ch >= 'a' && ch <= 'z' || ch == '_' || ch == '-');
```

Comment:
Should we be using a Regex vs. Linq here? Also this method should be called IsBase64UrlString.

Reply:
I think foreach is the fastest option here.

Reply:
Fixed